### PR TITLE
Add DialogRouter for Interactive Dialog to Apps Form migration

### PR DESCRIPTION
**Pull request by sbishel**
#### Summary
Implements feature flag-controlled routing between legacy InteractiveDialog and modern AppsForm components to enable gradual migration.

Mobile Implementation of
https://mattermost.atlassian.net/browse/MM-64700
https://github.com/mattermost/mattermost/pull/31821
 
Key Features:
- DialogRouter component with React.memo optimization for performance
- InteractiveDialogAdapter with WeakMap caching for form conversion
- Complete dialog/AppForm conversion utilities with validation
- Graceful fallback to legacy InteractiveDialog on conversion errors
- Type-safe implementation with optional subtype field support

Architecture:
- Feature flag controlled: FeatureFlagInteractiveDialogAppsForm
- Performance optimized: WeakMap cache, useMemo, useCallback patterns
- Error resilient: try/catch with fallback handling
- Mobile-first: Designed for React Native Navigation

Testing:
- 94 comprehensive unit tests covering all scenarios
- Cache behavior, error handling, edge cases
- Mock implementations for component integration
- Full TypeScript coverage with proper error cases

🤖 Generated with [Claude Code](https://claude.ai/code)

#### Ticket Link

  Fixes https://mattermost.atlassian.net/browse/MM-65649

#### Checklist

- [ x] Added or updated unit tests (required for all new features)
- [ ] Has UI changes
- [ ] Includes text changes and localization file updates
- [ ] Have tested against the 5 core themes to ensure consistency between them.
- [ ] Have run E2E tests by adding label `E2E iOS tests for PR`.

#### Device Information
This PR was tested on: 

#### Screenshots
None

#### Release Note

```release-note
Updated Interactive Dialog to use Apps Form Framework.
```

### CODE REVIEW

File: app/screens/dialog_router/dialog_router.tsx [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+    // Create performLookupCall - not used for basic dialogs but required by AppsFormComponent
+    const performLookupCall = useCallback(async (): Promise<DoAppCallResult<AppLookupResponse>> => {
+        return {data: {type: 'ok', data: {items: []}}};
+    }, []);
+
+    // Create refreshOnSelect - not used for basic dialogs but required by AppsFormComponent
+    const refreshOnSelect = useCallback(async (): Promise<DoAppCallResult<FormResponseData>> => {
+        return {data: {type: 'ok'}};
+    }, []);
```
Comment (sbishel) [inline app/screens/dialog_router/dialog_router.tsx:64]: These are placeholders and will be implemented with the PRs that require this functionality.

File: app/screens/apps_form/apps_form_field.tsx [hunk 2]
```diff
@@ -129,16 +133,21 @@ function AppsFormField({
     }, [field, value]);
 
     const selectedValue = useMemo(() => {
-        if (!value || !SelectableAppFieldTypes.includes(field.type || '')) {
+        if (!SelectableAppFieldTypes.includes(field.type || '')) {
             return undefined;
         }
 
-        if (!value) {
+        if (!value || value === '') {
```
Comment (larkox) [inline app/screens/apps_form/apps_form_field.tsx]: `!value` already implies `value === ''`, right?

**Comment by sbishel** [comment]
@saturninoabril - Mobile had no E2ETests for interactive dialog...I have added those but this may require some additional setup. A webhook server is required (detox/webhook_server.js) and we need to ensure that a feature flag (MM_FEATUREFLAGS_InteractiveDialogAppsForm) is set to True. These are my first mobile E2ETests...so I'm sure I may have missed something else.

**Comment by saturninoabril** [comment]
@sbishel Thanks for the heads up. I've created https://mattermost.atlassian.net/browse/MM-65711 (for configurable test server) and https://mattermost.atlassian.net/browse/MM-65712 (webhook server), @yasserfaraazkhan and I will discuss on how we can set this up separately.

**Comment by sbishel** [comment]
> **iOS E2E Test Report**: [d0f2ce7](https://github.com/mattermost/mattermost-mobile/commit/d0f2ce7768a22e46eda6b0906bf3b23719fdc448) | "82.21"% (268/326) | [full report](https://mattermost-detox-report.s3.amazonaws.com/17780929565-iPhone_15_Pro-iOS_17-2-d0f2ce7768a22e46eda6b0906bf3b23719fdc448-HEAD/jest-stare/ios-report.html)
> 
> Tests	Passed ✅	Failed ❌	Skipped ⏭️	Errors ⚠️
> 326	268	57	1	0

Am I missing something? I have a hard time believing this PR broke that many tests.

**Review by rahimrahman — APPROVED** [review]
A lot of unnecessary comments, which I imagine was done by AI.  If you're using Claude, there might be a way to add rules to avoid unnecessary comments.

**Comment by sbishel** [comment]
@saturninoabril, @yasserfaraazkhan -

I created new E2ETests to use a plugin, rather than webhook as requested here -
https://mattermost.atlassian.net/browse/MM-65712

I updated the plugin to add those endpoints here -
https://github.com/mattermost/mattermost-plugin-demo/pull/191

I know we need to add the plugin to the test server, but not sure how or where that's done. Let me know if we need to meet to move this forward. Thanks

**Comment by saturninoabril** [comment]
Nice, looking forward for the new feature of the demo plugin!

I suggest to handle it in the Detox itself. Globally, the non-prepackaged plugins are being disabled at `e2e/test/setup.ts` so it won't have side-effect to other tests. With that, it'd be good to upload/install then enable per spec file. Plugin APIs at `/detox/e2e/support/server_api/plugin.ts`. In monorepo, you'll find `cy.apiUploadAndEnablePlugin(demoPlugin)` where plugin can be installed via file or url which is a good approach to do the same here. Let me know if that helps, happy to jump in a call for any workarounds.

**Comment by sbishel** [comment]
@saturninoabril - Thanks for the pointers...I did just that...Once this PR for the demo plugin (https://github.com/mattermost/mattermost-plugin-demo/pull/191) gets merged and released. These tests will pass.

**Comment by yasserfaraazkhan** [comment]
work well locally. I think the demoplugin release is pending for e2e to pass on CI

**Comment by sbishel** [comment]
Dryrun security - Reliance on Client-Side Validation in app/utils/integrations.ts
- won't fix. These are strictly client side validation based on dialog configuration received from the plugin. It would be on the plugin to re-validate these on submission.

**Comment by sbishel** [comment]
@yasserfaraazkhan / @saturninoabril - I'm having a hard time getting the plugin installed and activated. Can I get access to the server (or different server) so I can see what's happening. I did it manually on a cloud server and everything worked.

One issue is we get a cloud flare timeout downloading the plug. I worked around that (i think) by downloaded the architecture specific tarball. Seems to work some of the time. However, it still won't activate. 

I'm open to suggestions. But we need to download and install and activate a plugin...I think i have the first 2 working...but can't get it activated. I'm guessing, config settings aren't sticking and activate is failing, but it times out instead.

**Comment by yasserfaraazkhan** [comment]
@sbishel 
1 job failed while installing SDK, but after rerunning it I can see the tests have pass.
image

**Comment by sbishel** [comment]
@yasserfaraazkhan - You apparently have the golden touch...it failed for me again in CI. I thought perhaps it was IOS vs Android...but both failed for me. They still work locally. It happens when attempting to activate the plugin. Although it still actually activates. I currently have a bunch of debug logging in the tests.
```
20:18:14.160 detox[8410] i user Enable failed with HTTP 524. Checking if plugin is actually active...
  origin: at e2e/support/server_api/plugin.ts:388:21
20:18:14.307 detox[8410] i user Plugin status after enable error: { isInstalled: true, isActive: true, version: '0.10.1' }
  origin: at e2e/support/server_api/plugin.ts:392:21
20:18:14.307 detox[8410] i user Plugin is actually active despite enable error. Treating as success.
  origin: at e2e/support/server_api/plugin.ts:399:25
20:18:14.307 detox[8410] i user Installation result: Plugin enabled successfully (despite HTTP 524 timeout)
  origin: at e2e/test/interactive_dialog/basic_dialog_plugin.e2e.ts:72:17
20:18:16.447 detox[8410] i user Final status: installed=true, active=true, version=0.10.1, versionMatch=false
```

  5. Root Cause
  The beforeAll hook is timing out because the plugin enable API call takes too long (hitting Cloudflare's timeout), causing the entire test setup to exceed the 180-second Jest timeout.

When I went to the machine directly and tried to activate manually. It was the same...180 seconds later got a 524...but the plugin did activate. 
Screenshot 2025-11-05 at 5 34 05 PM

I think the test is fine, but there is some issue with the environment or CloudFlare. Looking at the server logs, you can see the demo-plugin is running.

**Comment by yasserfaraazkhan** [comment]
Yes @sbishel I can see where its failing in the CI logs. [error ](https://github.com/mattermost/mattermost-mobile/actions/runs/19143234672/job/54716128151?pr=9114#step:14:1124)
`DEBUG  error on executeCommand Command with a trigger of 'dialog' not found. To send a message beginning with "/", try adding an empty space at the beginning of the message.`

On the Emulator this pop-up confirms the plugin is disabled. Although 2 of the tests pass initially.

testDone

I'll test a bit more on CI and see if there's a fix

File: detox/e2e/support/server_api/plugin.ts [hunk 3]
```diff
@@ -45,7 +78,7 @@ export const apiDisableNonPrepackagedPlugins = async (baseUrl: string): Promise<
         return;
     }
     plugins.active.forEach(async (plugin: any) => {
-        if (!prepackagedPlugins.has(plugin.id)) {
+        if (plugin.id !== DemoPlugin.id && !prepackagedPlugins.has(plugin.id)) {
```
Comment (sbishel) [inline detox/e2e/support/server_api/plugin.ts:80]: @yasserfaraazkhan - Since we are running them all at once, we don't want to keep this do we?

**Comment by sbishel** [comment]
@saturninoabril, @yasserfaraazkhan - All the detox tests have been very inconsistent. It seems like the longer the server is up...the more tests fail. 

I'm hesitant to check this in and add to the issue, but I also need to complete this PR and get it merged.

Thoughts?

**Comment by yasserfaraazkhan** [comment]
@sbishel there are 2 PR that have breaking changes 
https://github.com/mattermost/mattermost-mobile/pull/9270 causing failure in your pr. Before this merge, the tests on your Pr were at a good passing rate.

https://github.com/mattermost/mattermost-mobile/pull/9278 and this will break more due to keyboard changes. 

I have a PR up for fixing e2e https://github.com/mattermost/mattermost-mobile/pull/9371. 

I'll inform here once the fixes Pr merges. Probably in a day.

**Comment by sbishel** [comment]
Thanks @yasserfaraazkhan...that diagnosis was fast...nice.

**Comment by sbishel** [comment]
Dev Docs are not needed. This PR updates the architecture for the mobile Interactive Dialog...essentially catching up with the WebApp. No additional documentation is necessary.l
